SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,625 @@
+                    // changing how we model dynamic Self altogether.
+                    if callee.mayBindDynamicSelf {
+                        return paiToCallSiteDesc.intoCallSites(self.context)
+                    }
+
+                    // Proceed if the closure is passed as an argument (and not called).
+                    // If it is called we have nothing to do.
+                    //
+                    // `closureArgumentIndex` is the index of the closure in the callee's
+                    // argument list.
+                    guard let closureArgumentIndex = pai.calleeArgumentIndex(of: use) else {
+                        continue
+                    }    
+                    
+                    // Ok, we know that we can perform the optimization but not whether or
+                    // not the optimization is profitable. Check if the closure is actually
+                    // called in the callee (or in a function called by the callee).
+                    if !isClosureApplied(in: callee, index: closureArgumentIndex) {
+                        continue
+                    }
+
+                    // Gather the parameter info for the closure argument, at the apply site.
+                    guard let closureParamInfo = pai.operandConventions[parameter: use.index] else {
+                        fatalError("Parameter info not found for operand: \(use)!")
+                    }
+
+                    // We currently only support copying intermediate reabstraction
+                    // closures if the closure is ultimately passed trivially.
+                    let calleeConv = FunctionConvention(for: closureParamInfo.type, in: callee)
+                    let isClosurePassedTrivially = calleeConv.isTrivialNoescape
+
+                    // Mark the reabstraction closures as used.
+                    if haveUsedReabstraction {
+                        markReabstractionPartialApplyAsUsed(firstClosure: closureInst, currentClosure: use.value, usedReabstractionClosures: &usedReabstractionClosures)
+
+                        if !isClosurePassedTrivially {
+                            continue
+                        }
+                    }
+
+                    // Get all non-failure exit BBs in the Apply Callee if our partial apply
+                    // is guaranteed. If we do not understand one of the exit BBs, bail.
+                    //
+                    // We need this to make sure that we insert a release in the appropriate
+                    // locations to balance the +1 from the creation of the partial apply.
+                    //
+                    // However, thin_to_thick_function closures don't have a context and
+                    // don't need to be released.
+                    guard case Function.NonFailureExitBBsSearchResult.allBlocksUnderstood(let nonFailureExitBBs) = callee.findAllNonFailureExitBBs() else {
```

Comment:
formatting: Please try to stay within 120 chars per line

Reply:
Fixed.